### ISSUE TITLE
Integer overflow in ForsetiClient's heldCount leads to infinite loop during acquireExclusive()

### ISSUE THREAD

**Issue by tkroman** [neo4j/neo4j#6559]
- Neo4j version: 2.3.2
- Operating system: Ubuntu 14.04 x64
- API/Driver: Java API 2.3.2

TL;DR: here's a reproduction:

```
@Test
public void shouldAcquireExclusiveLock() throws Exception {
    Locks locks = dbRule
            .getGraphDatabaseAPI()
            .getDependencyResolver()
            .resolveDependency(Locks.class);

    Client client = locks.newClient();

    for (int i = 0; i < Integer.MAX_VALUE; i++) {
        client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L);
    } // heldCount = 7fffffff
    client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L); // heldCount = 80000000
    for (int i = 0; i < Integer.MAX_VALUE; i++) {
        client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L);
    } // heldCount = -1 now

    // and now WHOOOOPS... BAM!
    client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L);
}
```

When heldCount reaches -1 (after overflow, from below), we are forever stuck in this loop: https://github.com/neo4j/neo4j/blob/a58e9fd60c2c4aab06d4c2cba4e729f3e0c8a82d/enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java#L269

**Comment by tkroman** [comment]
OK, now it happened again. I don't know how frequent will it be, but now I have exact query that lead to this and some more debug stuff: 
query (via Java): 

```
db().execute(
  "MATCH (n:Node{treeId:{treeId}}) WITH n LIMIT 2500 DETACH DELETE n",
  MapUtil.map("treeId", treeId)
);
```

more screenshots (I'm really embarrassed about the way I do this :) ):
https://i.imgur.com/PktQAgC.png
https://i.imgur.com/KTSSrIV.png
https://i.imgur.com/yFkFxk8.png
https://i.imgur.com/NbWaAkp.png

Please let me know if there is something worth digging, so the next time it happens I can do more.

**Comment by chrisvest** [comment]
@tkroman It's hard to follow the screenshots. If you manage to boil it down to a more-or-less repeatable test case, then that would be very helpful.

It looks like the transaction is somehow deadlocking with itself, even though the locks are reentrant, which is what the `heldCount != -1` is about.

**Comment by tkroman** [comment]
@chrisvest, I'm with you on the subject of screenshot quality. I spent the whole day trying to write a unit-test, but it seems that no matter how many data I write, this  query works just fine. I'll try doing it overnight with random updates, and maybe execute the query manually tomorrow.
Here's what I know for now 
- those 2 cases arose in different apps
- one was during relationship creation (manually, Java API)
- another was during `DETACH DELETE`-int a node (essentially, the same case of attempting a lock on a `RELATIONSHIP` resource)

I'll keep trying to get something useful.

**Comment by tkroman** [comment]
@chrisvest, your comment about `heldCount != -1` was extremely helpful. Here's the reproduction:

```
@Test
public void shouldAcquireExclusiveLock() throws Exception {
    Locks locks = dbRule
            .getGraphDatabaseAPI()
            .getDependencyResolver()
            .resolveDependency(Locks.class);

    Client client = locks.newClient();

    for (int i = 0; i < Integer.MAX_VALUE; i++) {
        client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L);
    } // heldCount = 7fffffff
    client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L); // heldCount = 80000000
    for (int i = 0; i < Integer.MAX_VALUE; i++) {
        client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L);
    } // heldCount = -1 now

    // and now WHOOOOPS... BAM!
    client.acquireExclusive(ResourceTypes.RELATIONSHIP, 0L);
}
```

It hangs on the last line.

Basically, it's and integer overflow. When `heldCount` reaches -1 (after overflow, from below), we are forever stuck in this loop: https://github.com/neo4j/neo4j/blob/a58e9fd60c2c4aab06d4c2cba4e729f3e0c8a82d/enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java#L269

**Comment by chrisvest** [comment]
@tkroman This requires that you take the same lock (same node or same relationship, or same index entry) more than `Integer.MAX_VALUE` times in the same transaction. I find that a little improbable. What are you doing in your transactions, that would justify this conclusion?

**Comment by tkroman** [comment]
I didn't believe that myself, too. After that I checked the max. number of relationships connected to a single node: 

```
$match (n)-[r]-() return count(distinct r) as cnt order by cnt DESC limit 1
cnt 13149887 // that's 13*10^6
```

Is it possible that `DETACH DELETE` on such a node could lead to this issue?

**Comment by tkroman** [comment]
ping @chrisvest

**Comment by chrisvest** [comment]
It's on our backlog to look at this closer.

**Comment by chrisvest** [comment]
@tkroman No, a detach delete of a node with 13*10^6 relationships is not a problem here.

**Comment by tkroman** [comment]
@chrisvest, something strange is happening. I tried to use the `community` lock manager and it hang like this `Forseti` one. Under exactly the same conditions (taking a lock on a relationship, but the number of rels was significantly lesser (~hundreds)). Thread's been locked for hours (this is my local setup, so I can afford to not kill it):

```
at java.lang.Object.wait(Object.java:-1)
      at java.lang.Object.wait(Object.java:502)
      at org.neo4j.kernel.impl.locking.community.RWLock.acquireWriteLock(RWLock.java:388)
      at org.neo4j.kernel.impl.locking.community.LockManagerImpl.getWriteLock(LockManagerImpl.java:61)
      at org.neo4j.kernel.impl.locking.community.CommunityLockClient.acquireExclusive(CommunityLockClient.java:116)
      at org.neo4j.kernel.impl.api.LockingStatementOperations.relationshipCreate(LockingStatementOperations.java:282)
      at org.neo4j.kernel.impl.api.OperationsFacade.relationshipCreate(OperationsFacade.java:866)
      at org.neo4j.kernel.impl.core.NodeProxy.createRelationshipTo(NodeProxy.java:559)
```

**Comment by tkroman** [comment]
So, I was debugging this again and here's something even more _eerie_ than I expected:
https://i.imgur.com/A86x9Yk.png

Notice how the text under the selection (right screen) says
`// the threads in the waitingList cannot be currentThread`
And on the left half of the screen we see that `waitingList` is `[pool-6-thread-138]`, while `currentThread` IS `pool-6-thread-138`.

This happened after I left the app to spin for a while (3-4 hours), so I guess I may try leaving the same setup, but with `Forseti` client to check if I could dig something else.

**Comment by chrisvest** [comment]
@tkroman That you can reproduce this with the community lock manager means that the `heldCount` is definitely not to blame. Are you sure you don't have another transaction somewhere that neither commits or rolls back? A proper dead lock will kill one of the two transactions, but if there's nothing untoward with the lock ordering, a transaction may wait behind another for as long as the other transaction is alive. This is because we use pessimistic concurrency control.

**Comment by tkroman** [comment]
@chrisvest I believe I don't. Is it even possible for 2 transactions to behave this way in a single-thread bounds? And if that would be the case, it would reproduce in a (much) less flaky manner anyway. I'll keep digging.

P.S. I do think the bug with `heldCount` is still legit, though, since no matter how heavy a transaction is, no execution path should lead to an eternal while-loop.

**Comment by chrisvest** [comment]
@tkroman Just to confirm, your workload is entirely single-threaded – only a single transaction open at any time – and you can still reproduce this? If so, then this could point the finger at the object pooling we do internally.

**Comment by tkroman** [comment]
@chrisvest, it is single-threaded in a sense that the structure of the database is forest-like and we ensure that for each tree at each moment in time there's only one thread that is writing to it. 

I've reached a point where this bug reproduces in the fully set-up application environment with a probability of 1, but when I'm doing the same things in a unit-test, it just doesn't work. 

Maybe the difference is really in that there's more than one thread using lock clients or whatever. In that case I'll need to try and guess. But I checked - each transaction has it's `close()` called and there aren't even any nested transactions, so that it for now.

**Comment by chrisvest** [comment]
@tkroman Okay, and reach thread is writing to their respective tree in their own transaction, right?

**Comment by tkroman** [comment]
@chrisvest, yes of course, since transactions are thread-bound :)
Yee-haw, I've reproduced this in a stand-alone unit-test. I'll get rid of app-specific classes and will post it here in the next half an hour if everything is OK now.

**Comment by tkroman** [comment]
@chrisvest 
So, here's the deal. I had to twist the logic, so the code might not be sensible, but I hope it will give you the necessary perspective. I also attach a sample thread dump as seen from Idea's debugger (hangs during Cypher code execution). I couldn't reproduce it at first, had to make some changes here and there etc. However, I think that the fact that I managed to do this _without_ any custom code should ring some bells.

```
    @Test
    public void demo() throws Exception {
        String delQuery =
                "MATCH (n:Node { treeId:{treeId} }) " +
                "WHERE NOT n:Root " +
                "WITH n LIMIT 500 " +
                "DETACH DELETE n";
        Map<String, Object> qParams = MapUtil.map("treeId", 100L);
        Node root;
        Node dir;

        // create: '/'<--'tmp'<--'builder'<--'run'<-'dir'
        try (Transaction tx = db().beginTx()) {
            root = createTreeNode("/", null);
            root.addLabel(() -> "Root");
            Node tmp = createTreeNode("tmp", root);
            Node builder = createTreeNode("builder", tmp);
            Node run = createTreeNode("run", builder);
            dir = createTreeNode("dir", run);

            tx.success();
        }

        try (Transaction tx1 = db().beginTx()) {
            System.out.println(root.getAllProperties());
            tx1.success();
        }

        AtomicLong flushCounter = new AtomicLong(0L);
        Transaction tx = db().beginTx();
        for (int i = 0; i < 256; i++) {
            Node dirChild = createTreeNode(Integer.toHexString(i), dir);
            if (flushCounter.incrementAndGet() % 1000 == 0) {
                tx.success();
                tx.close();
                tx = db().beginTx();
            }
            for (int j = 0; j < 256; j++) {
                createTreeNode(Integer.toHexString(j), dirChild);
                if (flushCounter.incrementAndGet() % 1000 == 0) {
                    tx.success();
                    tx.close();
                    tx = db().beginTx();
                }
            }
        }
        tx.success();
        tx.close();

        try (Transaction tx1 = db().beginTx()) {
            System.out.println(root.getAllProperties());
            tx1.success();
        }

        while (true) {
            try (Transaction delTx = db().beginTx()) {
                Result deleted = db().execute(delQuery, qParams);
                if (deleted.getQueryStatistics().getNodesDeleted() == 0) {
                    break;
                }
                delTx.success();
            }
        }

        try (Transaction delTx = db().beginTx()) {
            for (Relationship r : root.getRelationships()) {
                r.delete();
            }
            root.delete();
            delTx.success();
        }
    }

    public Node createTreeNode(String name, @Nullable Node parent) {
        Node node = db().createNode(() -> "Node");
        node.setProperty("treeId", 100L);
        node.setProperty("fakeProp1", "qwerasdfzxcvzvcbsdfger{}xckvmlk{}lk;jfgksdlfg{}xvczxvsdfsafdsadf");
        node.setProperty("fakeProp2", 10L);
        node.setProperty("fakeProp3", false);

        if (parent != null) {
            node.createRelationshipTo(parent, () -> "CHILD_OF");
            node.setProperty("path", parent.getProperty("path") + "/" + name);
        } else {
            node.setProperty("path", name);
        }

        return node;
    }
```

Resulting thread dump should look like this:

```
"main@1" prio=5 tid=0x1 nid=NA sleeping
  java.lang.Thread.State: TIMED_WAITING
      at java.lang.Thread.sleep(Thread.java:-1)
      at java.lang.Thread.sleep(Thread.java:340)
      at org.neo4j.kernel.impl.util.concurrent.LockWaitStrategies$3.apply(LockWaitStrategies.java:69)
      at org.neo4j.kernel.impl.enterprise.lock.forseti.ForsetiClient.acquireExclusive(ForsetiClient.java:275)
      at org.neo4j.kernel.impl.api.LockingStatementOperations.relationshipDelete(LockingStatementOperations.java:307)
      at org.neo4j.kernel.impl.api.OperationsFacade.relationshipDelete(OperationsFacade.java:873)
      at org.neo4j.cypher.internal.spi.v2_3.TransactionBoundQueryContext$RelationshipOperations.delete(TransactionBoundQueryContext.scala:353)
      at org.neo4j.cypher.internal.spi.v2_3.TransactionBoundQueryContext$RelationshipOperations.delete(TransactionBoundQueryContext.scala:351)
      at org.neo4j.cypher.internal.compiler.v2_3.spi.DelegatingOperations.delete(DelegatingQueryContext.scala:165)
      at org.neo4j.cypher.internal.compatibility.ExceptionTranslatingQueryContextFor2_3$ExceptionTranslatingOperations.org$neo4j$cypher$internal$compatibility$ExceptionTranslatingQueryContextFor2_3$ExceptionTranslatingOperations$$super$delete(ExceptionTranslatingQueryContextFor2_3.scala:163)
      at org.neo4j.cypher.internal.compatibility.ExceptionTranslatingQueryContextFor2_3$ExceptionTranslatingOperations$$anonfun$delete$1.apply$mcV$sp(ExceptionTranslatingQueryContextFor2_3.scala:163)
      at org.neo4j.cypher.internal.compatibility.ExceptionTranslatingQueryContextFor2_3$ExceptionTranslatingOperations$$anonfun$delete$1.apply(ExceptionTranslatingQueryContextFor2_3.scala:163)
      at org.neo4j.cypher.internal.compatibility.ExceptionTranslatingQueryContextFor2_3$ExceptionTranslatingOperations$$anonfun$delete$1.apply(ExceptionTranslatingQueryContextFor2_3.scala:163)
      at org.neo4j.cypher.internal.compatibility.ExceptionTranslatingQueryContextFor2_3.org$neo4j$cypher$internal$compatibility$ExceptionTranslatingQueryContextFor2_3$$translateException(ExceptionTranslatingQueryContextFor2_3.scala:197)
      at org.neo4j.cypher.internal.compatibility.ExceptionTranslatingQueryContextFor2_3$ExceptionTranslatingOperations.delete(ExceptionTranslatingQueryContextFor2_3.scala:163)
      at org.neo4j.cypher.internal.compiler.v2_3.spi.UpdateCountingQueryContext$CountingOps.delete(UpdateCountingQueryContext.scala:154)
      at org.neo4j.cypher.internal.compiler.v2_3.mutation.DeleteEntityAction.org$neo4j$cypher$internal$compiler$v2_3$mutation$DeleteEntityAction$$delete(DeleteEntityAction.scala:58)
      at org.neo4j.cypher.internal.compiler.v2_3.mutation.DeleteEntityAction$$ano$$$$beb28fb77bc178c5befae9bc5fbf354$$$$eEntityAction$$delete$1.apply(DeleteEntityAction.scala:51)
      at org.neo4j.cypher.internal.compiler.v2_3.mutation.DeleteEntityAction$$ano$$$$beb28fb77bc178c5befae9bc5fbf354$$$$eEntityAction$$delete$1.apply(DeleteEntityAction.scala:51)
      at scala.collection.Iterator$class.foreach(Iterator.scala:742)
      at org.neo4j.cypher.internal.compiler.v2_3.helpers.BeansAPIRelationshipIterator.foreach(BeansAPIRelationshipIterator.scala:32)
      at org.neo4j.cypher.internal.compiler.v2_3.mutation.DeleteEntityAction.org$neo4j$cypher$internal$compiler$v2_3$mutation$DeleteEntityAction$$delete(DeleteEntityAction.scala:51)
      at org.neo4j.cypher.internal.compiler.v2_3.mutation.DeleteEntityAction.exec(DeleteEntityAction.scala:38)
      at org.neo4j.cypher.internal.compiler.v2_3.pipes.ExecuteUpdateCommandsPipe.org$neo4j$cypher$internal$compiler$v2_3$pipes$ExecuteUpdateCommandsPipe$$exec(ExecuteUpdateCommandsPipe.scala:55)
      at org.neo4j.cypher.internal.compiler.v2_3.pipes.ExecuteUpdateCommandsPi$$$$8751d7a9ee835394d5e1eaf9283dc3cd$$$$ands$1$$anonfun$apply$1.apply(ExecuteUpdateCommandsPipe.scala:44)
      at org.neo4j.cypher.internal.compiler.v2_3.pipes.ExecuteUpdateCommandsPi$$$$8751d7a9ee835394d5e1eaf9283dc3cd$$$$ands$1$$anonfun$apply$1.apply(ExecuteUpdateCommandsPipe.scala:44)
      at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:396)
      at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:396)
      at org.neo4j.cypher.internal.compiler.v2_3.pipes.EmptyResultPipe.internalCreateResults(EmptyResultPipe.scala:30)
      at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:125)
      at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:93)
      at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:63)
      at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:228)
      at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:212)
      at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anon$2.run(ExecutionPlanBuilder.scala:170)
      at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:200)
      at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:200)
      at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.runSafely(CompatibilityFor2_3.scala:116)
      at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper.run(CompatibilityFor2_3.scala:199)
      at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:27)
      at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:120)
      at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:114)
      at org.neo4j.cypher.javacompat.internal.ServerExecutionEngine.executeQuery(ServerExecutionEngine.java:62)
      at org.neo4j.kernel.impl.factory.GraphDatabaseFacade.execute(GraphDatabaseFacade.java:333)
      at ontrack.bush.util.Neo4jFeaturesTest.demo(Neo4jFeaturesTest.java:344)
      at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
      at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
      at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
      at java.lang.reflect.Method.invoke(Method.java:498)
      at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
      at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
      at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
      at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
      at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
      at org.junit.rules.RunRules.evaluate(RunRules.java:20)
      at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
      at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
      at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
      at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
      at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
      at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
      at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
      at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
      at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
      at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
      at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
      at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
      at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
      at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
      at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
```

**Comment by tkroman** [comment]
Also, should I split this thread into 2 separate properly (and succinctly) described issues, closing this one?

**Comment by chrisvest** [comment]
Yes, please. Make sure they refer back to this one, to keep history traceable.

**Comment by chrisvest** [comment]
@tkroman What OS/Hardware does this manifest on? I haven't seen the test fail after 100 trials on OS X.

**Comment by tkroman** [comment]
@chrisvest 
Hardware: Core i5-4570 @ 3.2GHz, 16GB RAM, some generic SSD, Ubuntu 14.04. 
JRE: 1.8.0_72-b15.
You can try to change these 1000, 500, 256 etc to some other values in comparable range, but this is very weird. I'm debugging this right now and it reproduces _every time_.

**Comment by tkroman** [comment]
So, I have 'printing breakpoints' on the following lines in `ForsetiClient`:

280 `heldLocks.put(resourceId, 1)`
657 `lockMap.remove(resourceId)`
680 `localLocks.put(resourceId, lockCount - 1)`
854 `releaseGlobalLock(lockMap, resourceId)`
541 `exclusiveLocks.clear()`
545 `exclusiveLockCounts[i] = Primitive.longIntMap()`

Now, thread hangs in `acquireExclusive` on `resourceId = 4974`.
Up to this point i got printed `put 4974`, which corresponds to line `280`, i.e. this resourceId is put into `heldLocks`  twice, but the thing is there is nothing printed about clearing between those two times. I can't fixate on some concrete resourceId since they are always different.

**Comment by chrisvest** [comment]
They should be cleared by `releaseAll` on commit/rollback.

**Comment by chrisvest** [comment]
I just tried on Windows with Java 1.8.0_45 and ArchLinux with Java 1.8.0_74, and both happily went through 100 invocations of the test.

**Comment by tkroman** [comment]
I don't know what to say. I've spent this whole day trying to find the reason. The only new thing I learned so far is that if I insert `ORDER BY id(n)` into the `WITH` clause, it doesn't hang.

**Comment by tkroman** [comment]
@chrisvest, I have to ask. Do you start an enterprise version of the database? The colleague of mine couldn't reproduce this as well until I noticed that he's running the community version. Also, do you run it in debug mode? Running it in ordinary (Run) mode may run OK for me now, too.

**Comment by tkroman** [comment]
@chrisvest, I'm so sorry for wasting your time. I used our test harness and totally forgot to mention that there are certain indexes created for this database. I created a project that reproduces this bug for me and all the people I can still reach in the office at 8PM.
The index that causes this issue is here: https://github.com/tkroman/demo/blob/master/src/main/java/Demo.java#L44. If you comment that line, everything works fine. Also, if you comment line 44 and un-comment line 43, everything works as well. Sometimes, however, it's worth to un-comment both lines to achieve the goal.
Can be run via Idea or by `mvn exec:java -Dexec.mainClass="Demo"`.
There's also an example of impact of `ORDER BY` present.

Again, sorry for wasting your time by being not clear on the details. Learned that lesson today :)
